### PR TITLE
Setup of new experimental workflow

### PR DESCRIPTION
**Pull request by dimstav23**
Done:
- Client/Controller/Server separation
- Parsing of the queries inside the controller itself
- Setup of new scripts for the client (`client.py`)
- Test for scalability w/ multithreaded workloads
- Implement termination of each client connection
- Adapt the automation scripts for the new experiments workflow (minor issue till we move the policy config on the client side - see [here](https://github.com/dimstav23/GDPRuler/issues/43))
- Send actual responses to the clients (not just ACK)
- Adapt the message buffer size for the logs - prepend the message with the size

### CODE REVIEW

File: GDPRuler.py [hunk 1]
```diff
@@ -39,21 +40,23 @@ def main():
...
-    process_args += ['--address', args.address]
+  if args.db_address:
+    process_args += ['--db_address', args.db_address]
   process_args += ['--logpath', args.logpath]
   if args.db_encryptionkey:
     process_args += ['--db_encryptionkey', args.db_encryptionkey]
   if args.log_encryptionkey:
     process_args += ['--log_encryptionkey', args.log_encryptionkey]
+  process_args += ['--frontend_address', args.frontend_address]
+  process_args += ['--frontend_port', args.frontend_port]
```
Comment (ertugrulaypek) [inline GDPRuler.py:51]: minor suggestion: could `controller_address` and `controller_port` be a bit more intiutive?

File: client.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  parser.add_argument('--address', help='IP address of the server to connect', default="127.0.0.1", required=False, type=str)
+  parser.add_argument('--port', help='Port of the running server to connect', default=1312, required=False, type=int)
+  parser.add_argument('--clients', help='Number of clients to spawn', default=1, type=int)
+  args = parser.parse_args()
+
+  # Start the time measurement before sending the workload
+  start_time = time.perf_counter()
+
+  latency_results = []
+  threads = []
```
Comment (ertugrulaypek) [inline client.py:62]: We can also consider using `multiprocessing` or` concurrent.futures.ProcessPoolExecutor` library as Python threading library can not utilize multiple cores due to GIL ([link](https://towardsdatascience.com/demystifying-python-multiprocessing-and-multithreading-9b62f9875a27)).

File: controller/source/controller.cpp [hunk 3]
```diff
@@ -161,6 +164,76 @@ auto handle_get_logs(const query &query_args,
...
+  std::array<char, max_msg_size> buffer{};
+  
+  // create the connection with the database instance
+  std::unique_ptr<kv_client> client = kv_factory::create(db_type, db_address);
+
+  while (true) {
+    // Read data from the socket
+    ssize_t bytes_read = recv(socket, buffer.data(), buffer.size() - 1, 0);
+    if (bytes_read <= 0) {
+      // Failed to read from socket or connection closed
```
Comment (ertugrulaypek) [inline controller/source/controller.cpp:179]: A log line can help us troubleshoot if an issue appears on connection.

**Review by ertugrulaypek — APPROVED** [review]

File: GDPRuler.py (hunk 1 above)
Comment (dimstav23) [reply GDPRuler.py:51]: Done

File: controller/source/controller.cpp (hunk 3 above)
Comment (dimstav23) [reply controller/source/controller.cpp:194]: added `std::cerr` output

File: client.py (hunk 2 above)
Comment (dimstav23) [reply client.py]: Legit and useful comment.
I ran some experiments w/o the connection, only with the scanning of the workloads and indeed multiprocessing scaled well while threading did not.
I adapted the client.

**Review by ertugrulaypek — APPROVED** [review]

**Comment by ertugrulaypek** [comment]
Hey @dimstav23, I tested them with the scenarios you mentioned and they all seem to be working :+1:

**Pull request merged**